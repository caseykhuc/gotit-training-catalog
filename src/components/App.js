/* eslint-disable import/no-named-as-default */
import React from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { Container } from 'react-bootstrap'
import PropTypes from 'prop-types';
import Header from './Header';

import { fetchCategory } from '../actions/category';
import { getCategories } from '../reducers';

import CategoryContainer from './Category/CategoryContainer';
import CategoryList from './Category/CategoryList';
import ItemNew from './Item/ItemNew';
import ModalContainer from './ModalContainer';


export class App extends React.Component {
  componentDidMount() {
    const { fetchCategory } = this.props;
    fetchCategory();
  }

  render() {
    const { categories } = this.props;

    return (
      <Container className="App">
        <Header />
        <CategoryList categories={categories} />
        <Switch>
          <Route path="/new-item" component={ItemNew} />
          <Route path="/categories/:categoryId" component={CategoryContainer} />
          {categories.length && (
            <Redirect to={`/categories/${categories[0].id}`} />
          )}
        </Switch>
        <ModalContainer />
      </Container>
    );
  }
}

const mapStateToProps = (state) => ({
  categories: getCategories(state),
});


App.propTypes = {
  categories: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
  })).isRequired,
  fetchCategory: PropTypes.func,
}

export default connect(mapStateToProps, { fetchCategory })(App);
