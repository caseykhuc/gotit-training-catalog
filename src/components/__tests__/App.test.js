import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps } from '../App';

describe('component/App', () => {
  let props;
  let wrapper;
  const update = () => {
    wrapper.update();
  };
  const setup = () => {
    wrapper = shallow(<App {...props} />);
    update();
  };
  beforeEach(() => {
    props = {
      categories: [{
        id: 1,
        name: 'Cat',
      }],
      fetchCategory: jest.fn(),
      fetchUser: jest.fn(),
      isLoading: false,
    }
  });
  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();

    props.categories = [];
    setup();
    expect(wrapper).toMatchSnapshot();

    props.isLoading = true;
    setup();
    expect(wrapper).toMatchSnapshot();
  });
  it('should fetch  user when mounted', () => {
    setup();
    expect(props.fetchUser).toHaveBeenCalled();
  })
})

describe('component/App (mapStateToProps)', () => {
  let state;
  beforeEach(() => {
    state = {
      category: {
        byId: {},
        isLoading: false,
        error: 'Category Error',
      },
      user: {
        isLoading: true,
        error: 'User Error',
        userId: 1,
      },
    };
  });

  it('should return isLoading/error values properly', () => {
    expect(mapStateToProps(state).isLoading).toBe(true);
    expect(mapStateToProps(state).error).toBe('Category Error');

    state.user.isLoading = false;
    expect(mapStateToProps(state).isLoading).toBe(false);
  });
});
