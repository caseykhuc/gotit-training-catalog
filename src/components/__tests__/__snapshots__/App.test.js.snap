// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Category/CategoryContainer should render correctly 1`] = `ShallowWrapper {}`;

exports[`component/Category/CategoryContainer should render correctly 2`] = `ShallowWrapper {}`;
